File: packages/dotnet/NaturalSpacing.Core/PolicyAndText.cs
using System.Text.RegularExpressions;

namespace NaturalSpacing.Core;

public static class NaturalSpacingPolicy
{
    private static readonly HashSet<ContentKind> NaturalKinds =
        [ContentKind.Prose, ContentKind.Title, ContentKind.Message, ContentKind.Note,
         ContentKind.Document, ContentKind.Transcript, ContentKind.AsrTranscript];
    private static readonly HashSet<ContentKind> VerbatimKinds =
        [ContentKind.Code, ContentKind.Identifier, ContentKind.Url, ContentKind.Email,
         ContentKind.Password, ContentKind.Token, ContentKind.FilePath, ContentKind.Command, ContentKind.Number];
    private static readonly Regex[] StructuredPatterns =
    [
        new(@"^(?:[a-z][a-z0-9+.-]*://|www\.)", RegexOptions.IgnoreCase),
        new(@"^[^\s@]+@[^\s@]+\.[^\s@]+$", RegexOptions.IgnoreCase),
        new(@"^(?:[A-Za-z]:[\\/]|~?[\\/])"),
        new(@"^(?:[A-Za-z_$][A-Za-z0-9_$-]*[.:/@\\])+[A-Za-z0-9_$.-]+$"),
        new(@"(?:=>|::|</?[A-Za-z][^>]*>|\{[^}]*}|`[^`]*`)"),
    ];

    public static PolicyRecommendation Recommend(PolicyContext? context = null)
    {
        context ??= new PolicyContext();
        if (context.IsSecure == true || context.ContentKind == ContentKind.Password)
            return Result(FieldPolicy.Verbatim, RecommendationConfidence.High, RecommendationSource.Safety, RecommendationReason.SecureContent, true);
        if (context.ExplicitPolicy is { } selectedPolicy)
            return Result(selectedPolicy, RecommendationConfidence.High, RecommendationSource.Explicit, RecommendationReason.ExplicitPolicy, true);
        var kind = context.ContentKind ?? ContentKind.Unknown;
        if (NaturalKinds.Contains(kind))
            return Result(FieldPolicy.NaturalLanguage, RecommendationConfidence.High, RecommendationSource.ContentKind, RecommendationReason.NaturalLanguageContent, true);
        if (VerbatimKinds.Contains(kind))
            return Result(FieldPolicy.Verbatim, RecommendationConfidence.High, RecommendationSource.ContentKind, RecommendationReason.StructuredContent, true);
        var value = context.Text?.Trim();
        if (!string.IsNullOrEmpty(value) && StructuredPatterns.Any(pattern => pattern.IsMatch(value)))
            return Result(FieldPolicy.Verbatim, RecommendationConfidence.Medium, RecommendationSource.TextHeuristic, RecommendationReason.StructuredText, false);
        if (kind == ContentKind.SearchQuery)
            return Result(FieldPolicy.NaturalLanguage, RecommendationConfidence.Medium, RecommendationSource.ContentKind, RecommendationReason.AmbiguousSearch, false);
        if (context.Text is not null && NaturalSpacingFormatter.EligibleInsertions(context.Text).Count > 0)
            return Result(FieldPolicy.NaturalLanguage, RecommendationConfidence.Medium, RecommendationSource.TextHeuristic, RecommendationReason.MixedNaturalLanguage, false);
        return Result(FieldPolicy.Verbatim, RecommendationConfidence.Low, RecommendationSource.Fallback, RecommendationReason.InsufficientEvidence, false);
    }

    public static FieldPolicy Resolve(
        PolicyContext? context = null,
        FieldPolicy fallback = FieldPolicy.Verbatim)
    {
        var recommended = Recommend(context);
        return recommended.AutoApply ? recommended.Policy : fallback;
    }

    public static FormattedTextUpdate Format(TextUpdate update)
    {
        var display = NaturalSpacingFormatter.Normalize(update.Text, update.Policy);
        return new FormattedTextUpdate(display, update.Stability == TextStability.Final ? display : null,
            display != update.Text, update.Policy, update.Source, update.Stability);
    }

    private static PolicyRecommendation Result(FieldPolicy policy, RecommendationConfidence confidence,
        RecommendationSource source, RecommendationReason reason, bool autoApply) =>
        new(policy, confidence, source, reason, autoApply);
}

File: packages/kotlin/src/main/kotlin/dev/naturalspacing/core/PolicyAndText.kt
package dev.naturalspacing.core

private val naturalKinds = setOf(
    ContentKind.PROSE, ContentKind.TITLE, ContentKind.MESSAGE, ContentKind.NOTE,
    ContentKind.DOCUMENT, ContentKind.TRANSCRIPT, ContentKind.ASR_TRANSCRIPT,
)
private val verbatimKinds = setOf(
    ContentKind.CODE, ContentKind.IDENTIFIER, ContentKind.URL, ContentKind.EMAIL,
    ContentKind.PASSWORD, ContentKind.TOKEN, ContentKind.FILE_PATH, ContentKind.COMMAND,
    ContentKind.NUMBER,
)
private val structuredPatterns = listOf(
    Regex("^(?:[a-z][a-z0-9+.-]*://|www\\.)", RegexOption.IGNORE_CASE),
    Regex("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$", RegexOption.IGNORE_CASE),
    Regex("^(?:[A-Za-z]:[\\\\/]|~?[\\\\/])"),
    Regex("^(?:[A-Za-z_$][A-Za-z0-9_$-]*[.:/@\\\\])+[A-Za-z0-9_$.-]+$"),
    Regex("(?:=>|::|</?[A-Za-z][^>]*>|\\{[^}]*}|`[^`]*`)"),
)

fun recommendPolicy(context: PolicyContext = PolicyContext()): PolicyRecommendation {
    if (context.isSecure == true || context.contentKind == ContentKind.PASSWORD) {
        return recommendation(FieldPolicy.VERBATIM, RecommendationConfidence.HIGH, RecommendationSource.SAFETY, RecommendationReason.SECURE_CONTENT, true)
    }
    context.explicitPolicy?.let {
        return recommendation(it, RecommendationConfidence.HIGH, RecommendationSource.EXPLICIT, RecommendationReason.EXPLICIT_POLICY, true)
    }
    val kind = context.contentKind ?: ContentKind.UNKNOWN
    if (kind in naturalKinds) return recommendation(FieldPolicy.NATURAL_LANGUAGE, RecommendationConfidence.HIGH, RecommendationSource.CONTENT_KIND, RecommendationReason.NATURAL_LANGUAGE_CONTENT, true)
    if (kind in verbatimKinds) return recommendation(FieldPolicy.VERBATIM, RecommendationConfidence.HIGH, RecommendationSource.CONTENT_KIND, RecommendationReason.STRUCTURED_CONTENT, true)
    if (context.text?.trim()?.let { value -> value.isNotEmpty() && structuredPatterns.any { it.containsMatchIn(value) } } == true) {
        return recommendation(FieldPolicy.VERBATIM, RecommendationConfidence.MEDIUM, RecommendationSource.TEXT_HEURISTIC, RecommendationReason.STRUCTURED_TEXT, false)
    }
    if (kind == ContentKind.SEARCH_QUERY) return recommendation(FieldPolicy.NATURAL_LANGUAGE, RecommendationConfidence.MEDIUM, RecommendationSource.CONTENT_KIND, RecommendationReason.AMBIGUOUS_SEARCH, false)
    if (context.text?.let { NaturalSpacing.eligibleInsertions(it).isNotEmpty() } == true) {
        return recommendation(FieldPolicy.NATURAL_LANGUAGE, RecommendationConfidence.MEDIUM, RecommendationSource.TEXT_HEURISTIC, RecommendationReason.MIXED_NATURAL_LANGUAGE, false)
    }
    return recommendation(FieldPolicy.VERBATIM, RecommendationConfidence.LOW, RecommendationSource.FALLBACK, RecommendationReason.INSUFFICIENT_EVIDENCE, false)
}

fun resolvePolicy(
    context: PolicyContext = PolicyContext(),
    fallback: FieldPolicy = FieldPolicy.VERBATIM,
): FieldPolicy {
    val recommended = recommendPolicy(context)
    return if (recommended.autoApply) recommended.policy else fallback
}

fun formatTextUpdate(update: TextUpdate): FormattedTextUpdate {
    val display = NaturalSpacing.normalize(update.text, update.policy)
    return FormattedTextUpdate(
        display,
        if (update.stability == TextStability.FINAL) display else null,
        display != update.text,
        update.policy,
        update.source,
        update.stability,
    )
}

private fun recommendation(
    policy: FieldPolicy,
    confidence: RecommendationConfidence,
    source: RecommendationSource,
    reason: RecommendationReason,
    autoApply: Boolean,
) = PolicyRecommendation(policy, confidence, source, reason, autoApply)
